agot-bg-game-server/src/common/ingame-game-state/action-game-state/resolve-march-order-game-state/combat-game-state/post-combat-game-state/resolve-retreat-game-state/ResolveRetreatGameState.ts

```diff
@@ -73,7 +74,12 @@ export default class ResolveRetreatGameState extends GameState<
             // The loser is the attacker...
             if (overridenChooser == null) {
                 // ...and the chooser was not overriden:
-                // The units retreat automatically to the region they came from.
+                // The units retreat automatically to the region they came from unless they attacked from a enemy capital and haven't left a power token
+                if(this.attackingUnitsCannotRetreat(this.combat.attackingRegion, this.postCombat.loser)) {
```

Comment:
You can simply check if the `attackingRegion` is controlled by an other house than `attacker` at the moment the retreat must be made. If it isn't then, they must all be killed.

Reply:
I don't think that will help as `getController()` will return the attacker as his units are still present in `attackingRegion` at this point.

Reply:
And checking `superControlPowerToken`?

On top of that, I think if player A moves from the capital of player B to attack player C without leaving a Power token and loses, the code at the moment won't correctly kill all units.

Reply:
> 
> 
> And checking `superControlPowerToken`?
exactly that's done in attackingUnitsCannotRetreat (which will be renamed to attackingRegionBlockedForRetreat() to support Robb) but I decided to do the checks in method as this is more readable than a complex if chain.

> 
> On top of that, I think if player A moves from the capital of player B to attack player C without leaving a Power token and loses, the code at the moment won't correctly kill all units.

Why?

Reply:
I fixed it that Robb Stark won't be able to retreat to the startingRegion. In addition I modified the unit-loss-warning to support all possible cases though ship loss will come with a separate PR but then I remove the logic from there.